### PR TITLE
Award badges for github commits with various milestones

### PR DESCRIPTION
**Pull request by vishaldeepak**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Solves issue #4002 . We now award badge awards from Github at various stages. 1,16,32 and 64. They have been name `dev-contributor`(as it is currently), `dev-contributor-bronze`, `dev-contributor-silver`, `dev-contributor-gold`

## Related Tickets & Documents
#4002 

## QA Instructions, Screenshots, Recordings

Changes in file include setting up a new constant Badge_SLUGS. We now send 1 request each to fetch all the commits from github for a particular repo. We then group all commits as key value pairs of `id => commits_count`. 

## Added/updated tests?

- [x] Yes

### CODE REVIEW

File: app/services/badges/award_contributor_from_github.rb
Comment (vishaldeepak) [inline app/services/badges/award_contributor_from_github.rb:52]: Possibility of a lot of queries firing here.

File: app/services/badges/award_contributor_from_github.rb
Comment (djuber) [inline app/services/badges/award_contributor_from_github.rb:27]: One efficiency concern - this used to return just the last 1 days worth of commits for each repo, now it returns a long list of all commits (which are in pages of 100, and we have over 9000 commits in the forem/forem repo).

Assuming the rate limiting is not an issue (it was for me, locally, since I had not setup github authentication tokens in my .env file, it looks like it was active for you as well based on the VCR cassette) there is still a question of space and time used by us locally.

```ruby
[6] pry(main)> Benchmark.measure { commits = Github::OauthClient.new.commits(repo, since: 1.day.ago) }
=> #<Benchmark::Tms:0x0000558876410970
 @cstime=0.0,
 @cutime=0.0,
 @label="",
 @real=0.3045605589868501,
 @stime=0.007760000000000003,
 @total=0.05495700000000005,
 @utime=0.047197000000000044>
```

This was to get a page of 7 commits from the last day - but it took about 300ms.

I tried to ask for the full forem/forem repo to confirm the behavior - and saw a series of paged requests, one by one, until I had hit the 60 requests per hour api rate limit for unauthenticated requests. I got 5700 out of 9546 commits back (the client handled the 403 response by returning the accumulated results, which were most recent first). 

I ran it again after the rate limit expired, making 60 requests took 49.7 seconds, and that's only 2/3rds of the time needed - so this job would require about 1m20s to pull the repositories data down.

We currently run this on a schedule every hour (in addition to any other github api requests we make elsewhere, like Readme preview rendering for linked  repositories) so there's a possible chance that as the number of commits or repos grows, the authenticated rate limit (5000 requests per hour) may be exceeded. Right now we would expect only 100-200 requests from this service when it runs (total number of commits from four repositories) but if we added public repositories to track for awards, or the number of commits continued to grow, this might present some limitations later.

It would be much nicer from an execution time and memory use standpoint if we could limit the request to a window (like the 1.day.ago filter that was here) and increment a repo:commits counter for the user when a new commit is seen, that probably adds some complexity, or use another endpoint (`stats/contributors`, Octokit exposes this as `contributor_stats(repo)`) 

https://docs.github.com/en/rest/reference/repos#get-all-contributor-commit-activity documents this, `total` and `author.id` look like they are the relevant items we want and are equivalent to fetching all commits, and counting by author id. 

It might be worthwhile to validate the counts are the same or comparable between the two endpoints, but this appears to be a single request with an aggregated response (or a smaller set of requests) already grouped by github user, if we're planning to compare cumulative commit counts rather than looking for new committers (as we had done previously).

File: app/services/badges/award_contributor_from_github.rb
Comment (djuber) [reply app/services/badges/award_contributor_from_github.rb:5]: answered in a subsequent comment - there's an existence check further down.

**Review by djuber — COMMENTED** [review]
I think we want to investigate the best way to get this information (github user id, number of commits to the repo) more efficiently than pulling all commits from all time. 

Either we need to keep track of "seen" commits per user and update using new commits, or see if using the /stats/contributors api endpoint gives us the same number we're looking to add up ourselves here.

Most of the other comments were just reading notes and are probably less critical than making the right choice about retrieving this information.

File: app/services/badges/award_contributor_from_github.rb
Comment (djuber) [reply app/services/badges/award_contributor_from_github.rb:27]: I tested the contributors stats, and I get a single page (of 100) -  this appears to be by design https://docs.github.com/en/github/visualizing-repository-data-with-graphs/accessing-basic-repository-data/viewing-a-projects-contributors

Right now this limits the view to contributors with four or more commits to the repo (in forem/forem) and is probably going to miss all first time contributors. Would it make sense to pass through this twice per repo, once with the last day's new commits, for first time contributors, and again using the statistics endpoint for cumulative badges?

Comment (djuber) [reply app/services/badges/award_contributor_from_github.rb:27]: Doing a little digging - it looks like passing the anon=true option (`contributors("forem/forem", true)` in Octokit's client) gives 618 contributors rather than the last 100. 

There were 178 nil ids here, but this includes the long tail of users with a single commit, and only takes 7 requests (paginated by 100). 

I propose we rework the request to use this instead of commits(repo):

```ruby
Github::OauthClient.new.contributors("forem/forem", true) # anon=true for full contributor list instead of summary graph
```

The response format changes (so this is a slightly larger rework to the api response handling than I would have liked), we're dealing with an array of objects like this:

```
{:login=>"dependabot[bot]",
 :id=>49699333,
 :node_id=>"MDM6Qm90NDk2OTkzMzM=",
 :avatar_url=>"https://avatars.githubusercontent.com/in/29110?v=4",
 :gravatar_id=>"",
 :url=>"https://api.github.com/users/dependabot%5Bbot%5D",
 :html_url=>"https://github.com/apps/dependabot",
 :followers_url=>"https://api.github.com/users/dependabot%5Bbot%5D/followers",
 :following_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/following{/other_user}",
 :gists_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/gists{/gist_id}",
 :starred_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/starred{/owner}{/repo}",
 :subscriptions_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/subscriptions",
 :organizations_url=>"https://api.github.com/users/dependabot%5Bbot%5D/orgs",
 :repos_url=>"https://api.github.com/users/dependabot%5Bbot%5D/repos",
 :events_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/events{/privacy}",
 :received_events_url=>
  "https://api.github.com/users/dependabot%5Bbot%5D/received_events",
 :type=>"Bot",
 :site_admin=>false,
 :contributions=>1093}
```

I assume `id` and `contributions` are the key factors needed for the result here, the results are in descending order of contributions, and the sum of the returned contributions is equal to the number of commits shown on the repo page in github so it included everything.

Comment (vishaldeepak) [reply app/services/badges/award_contributor_from_github.rb:27]: sounds good, let me take a look at this

File: app/services/badges/award_contributor_from_github.rb
Comment (vishaldeepak) [reply app/services/badges/award_contributor_from_github.rb:51]: @djuber Love the idea, my only concern here is what happens if there are many users who have reached a bronze level. Lets assume there are 100 users that have bronze(4 commits). So this would mean 100 queries to the database. Sounds reasonable?
Thats the only reason I went for memoize

File: app/services/badges/award_contributor_from_github.rb
Comment (vishaldeepak) [reply app/services/badges/award_contributor_from_github.rb]: @djuber Made changes as requested.

File: app/services/badges/award_contributor_from_github.rb
Comment (vishaldeepak) [reply app/services/badges/award_contributor_from_github.rb:4]: Yeah I think its better if I team member is on it.

Comment (citizen428) [reply app/services/badges/award_contributor_from_github.rb:4]: Ok, no problem we can do it after this PR has been merged.

**Comment by Zhao-Andy** [comment]
Just wanted to add that new badges will always need new badge images to go along with it.

**Comment by msarit** [comment]
Hey fellas! @djuber @citizen428 @Zhao-Andy

What are your thoughts on this PR? Is this something the author should move forward with?

**Comment by benhalpern** [comment]
Hey @emmmmmmmmmmmma that sounds great!

To @citizen428's point we really should transition these thematically to Forem (I was unsure of whether we should preserve the historical context)...

I will follow up tomorrow on some light creative direction and then let you take it wherever you want.

This is going to be really fun to close the loop on, glad we're doing this!

**Comment by citizen428** [comment]
I'm not approving this just yet because we're still waiting for icons and I don't want this PR to accidentally get merged.

**Comment by msarit** [comment]
@emmmmmmmmmmmma Hello! 👋🏾 
Any update on the icons for this PR?

**Comment by emmmmmmmmmmmma** [comment]
Hey @msarit ! Apologies for the wait. [I have a set of badges here that can be used for this here, with slight color variations.](https://drive.google.com/drive/folders/1LacYT-s-bTL8krGKZ6T5RkvfeBOlB6xH?usp=sharing) Let me know if you have feedback or need anything else!

**Comment by msarit** [comment]
@vishaldeepak Hello there!
The icon designs are now available; check Emma’s comment above.
Let us know if you need support or clarification 🙏🏾

**Comment by vishaldeepak** [comment]
I'l take a look, thank you.

**Comment by citizen428** [comment]
Hi @vishaldeepak Just checking in to see if you'll have time to finish this PR or if a core team member should take over from here?

**Comment by vishaldeepak** [comment]
@citizen428 Apologies, totaly overlooked this.  Looks like I cannot access the above google drive link.

**Comment by citizen428** [comment]
@vishaldeepak I turned them into a zip file for you: [badges-20220103T062348Z-001.zip](https://github.com/forem/forem/files/7800630/badges-20220103T062348Z-001.zip)

**Comment by vishaldeepak** [comment]
@citizen428 Hey I thinking I am missing some context. 
- Which image maps to which badge? . I dont have access to comment posted by @emmmmmmmmmmmma 
- How do versions for images fit to the current task? 
- Looks like the task now is to create new Badge via seed file or perhaps some rake task and attach appropriate images to these badges. For the current dev badge given out which image do we use? Is it present in the `app/assets/images` ?

**Comment by aitchiss** [comment]
Hey @vishaldeepak - sorry for the delay but just to let you know I'm following up on your questions and will update you soon! Thanks for bearing with us - I'm excited to see these new badges get implemented 🤩

**Comment by JennieOcken** [comment]
@vishaldeepak 
Jumping in to provide decisions.  If these don't make sense, please push back on me... And then ask for technical direction.
> * Which image maps to which badge? 
1: Branch
8: Cycle
16: Sprout
32: Heart
64: Wink

(I believe we didn't have 8 in here before but I am hopeful it is easy to add)

> * How do versions for images fit to the current task?
> * Looks like the task now is to create new Badge via seed file or perhaps some rake task and attach appropriate images to these badges. For the current dev badge given out which image do we use? Is it present in the `app/assets/images` ?

I don't know.  @citizen428 help!?  I don't know how badges work 😸 

Thank you!

**Comment by citizen428** [comment]
1. Badges are managed in the DB. The model is `Badge`, the table `badges`.
2. The name of the image is stored in the `badge_image` column. Here's the result of the following query: `select title, slug, description, badge_image from badges where slug like '%streak'` as inspiration/template:
    Screen Shot 2022-02-01 at 10 15 14
3. This means we'd have to add a [data update script](https://developers.forem.com/backend/data-update-scripts) for generating badge entries in this table for existing Forems. I'm not sure if we set up badges for new Forems at all, I can't find a rake tasks for this. Maybe @maestromac knows?
4. The `Badge` class mounts a CarrierWave uploader, so badges are stored on S3 it seems. @Zhao-Andy probably knows more about this and may also be able to help with uploading the images.

**Comment by vishaldeepak** [comment]
Hey guys, 
I have added a new commit and rebased with main. 
The new commit contains a script to add badges. Since I don't have access to add images, the commit is bit limiting. 

I think any core-team member should be able to proceed now. Upload appropriate images and everything should work as expected.

**Comment by citizen428** [comment]
@vishaldeepak Thank you so much for both this contribution as well as your patience getting this as close to the finish line as possible as an external contributor. ❤️ 

@maestromac Can you help with finalizing this task?

**Review by citizen428 — APPROVED** [review]

**Comment by maestromac** [comment]
:wave: @vishaldeepak , I'm going to take over this and help push it over the finish line. There are still some design decisions that we need to consider, so I'm going to remove the DUS script and a few other lines of code.

File: spec/services/badges/award_contributor_from_github_spec.rb
Comment (maestromac) [inline spec/services/badges/award_contributor_from_github_spec.rb:13]: use the same cassette for all the specs for optimization purposes.

File: app/services/badges/award_contributor_from_github.rb
Comment (maestromac) [inline app/services/badges/award_contributor_from_github.rb:29]: After comparing award single vs award multi a bit, I realize that they are both rewarding the base `dev-contributor` which is redundant work. award multi is enough to do all the work, unless I missed something @vishaldeepak ?

Comment (vishaldeepak) [reply app/services/badges/award_contributor_from_github.rb:29]: @maestromac  This should answer your question. https://github.com/forem/forem/pull/14381#discussion_r683720317

Comment (maestromac) [reply app/services/badges/award_contributor_from_github.rb]: After speaking with @djuber , I decided to undo my changes. There's limitation to the contributor API so it might be a better idea to keep it as is.

**Review by djuber — APPROVED** [review]

**Comment by maestromac** [comment]
Thank you for all your work @vishaldeepak !

**Pull request merged**
